Source/WebInspectorUI/UserInterface/Controllers/DeviceSettingsManager.js

```diff
@@ -0,0 +1,194 @@
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+WI.DeviceSettingsManager = class DeviceSettingsManager extends WI.Object
+{
+    constructor()
+    {
+        super();
+
+        this._deviceUserAgent = null;
+        this._overridenDeviceUserAgent = null;
+        this._overridenDeviceScreenSize = null;
+        this._overridenDeviceSettings = new Map;
+    }
+
+    // Target
+
+    initializeTarget(target)
+    {
+        if (!target.hasDomain("Page"))
+            return
+
+        // COMPATIBILITY (iOS 12.2): Page.overrideUserAgent did not exist.
+        if (target.hasCommand("Page.overrideUserAgent") && this._overridenDeviceUserAgent)
+            target.PageAgent.overrideUserAgent(this._overridenDeviceUserAgent);
+
+        // COMPATIBILITY (iOS 12.2): Page.overrideSetting did not exist.
+        if (target.hasCommand("Page.overrideSetting")) {
+            for (let [setting, value] of this._overridenDeviceSettings)
+                target.PageAgent.overrideSetting(setting, value);
+        }
+
+        const options = {
+            objectGroup: WI.DeviceSettingsManager.ObjectGroup,
+            returnByValue: true,
+        }
+
+        WI.runtimeManager.evaluateInInspectedWindow("globalThis.navigator?.userAgent", options, (remoteObject, wasThrown, result) => {
+            if (!wasThrown && result)
+                this._deviceUserAgent = result.value;
+
+            target.RuntimeAgent.releaseObjectGroup(WI.DeviceSettingsManager.ObjectGroup);
+        });
+    }
+
+    // Static
+
+    static supportsSetScreenSizeOverride()
+    {
+        return InspectorBackend.hasCommand("Page.setScreenSizeOverride");
+    }
+
+    // Public
+
+    get deviceUserAgent() { return this._deviceUserAgent; }
+    get overridenDeviceUserAgent() { return this._overridenDeviceUserAgent; }
+    get overridenDeviceScreenSize() { return this._overridenDeviceScreenSize; }
+    get overridenDeviceSettings() { return this._overridenDeviceSettings; }
+
+    get hasOverridenDefaultSettings() {
+        return this._overridenDeviceUserAgent || this._overridenDeviceScreenSize || this._overridenDeviceSettings.size > 0;
+    }
+
+    overrideDeviceSetting(setting, value, callback)
+    {
+        let target = WI.assumingMainTarget();
+        if (!target.hasCommand("Page.overrideSetting")) {
+            callback(false);
+            return;
+        }
+
+        let commandArguments = {
+            setting,
+        };
+
+        let apply = false;
+        if (!this._overridenDeviceSettings.has(setting)) {
+            apply = true;
+            commandArguments.value = value; // Providing `value` applies the override for `setting`. Not providing it clears the override.
+        }
```

Comment:
Nit:
```
let apply = !this._overridenDeviceSettings.has(setting);
if (apply)
    commandArguments.value = value;
```

Also can we provide a slightly more descriptive name for `apply`? Maybe `shouldApplyValue` or maybe invert the setting and call it `shouldRemoveOverride`? A clearer name would likely eliminate the need for this comment as well. This applies to the methods below as well that have an `apply` used this way.

Reply:
Sure, we can use a more descriptive name, like `shouldApplyValue`.

But IMO we should keep the comment. To figure out that _not passing a value_ removes the override isn't immediately obvious. One needs to either read the description of the `overrideSetting` entry in `Page.json` or read through the implementation of `InspectorPageAgent.cpp`.